source/fundamentals/gridfs.txt

```diff
@@ -4,7 +4,353 @@ GridFS
+bucket. Each document in the ``files`` collection contains the following
+information:
+
+- The file ID
+- The file length
+- The maximum chunk size
+- The upload date and time
+- The file name
+- A ``metadata`` document in which you can store any other information
+
+To retrieve file data, call the ``Find()`` method on a ``GridFSBucket``
+instance. You can pass a query filter as an argument to ``Find()`` to match
+only certain file documents.
+
+.. note::
+
+   The ``Find()`` method requires a query filter as a parameter. To match all 
+   documents in the ``files`` collection, pass an empty query filter to ``Find()``.
+
+The following example retrieves the file name and length of documents in the
+``files`` collection with ``length`` values greater than ``1500``:
+
+.. code-block:: go
+
+   filter := bson.D{{"length", bson.D{{"$gt", 1500}}}}
+   cursor, err := bucket.Find(filter)
+   if err != nil {
+      panic(err)
+   }
+
+   type gridfsFile struct {
+      Name   string `bson:"filename"`
+      Length int64  `bson:"length"`
+   }
+   var foundFiles []gridfsFile
+   if err = cursor.All(context.TODO(), &foundFiles); err != nil {
+      panic(err)
+   }
+
+   for _, file := range foundFiles {
+      fmt.Printf("filename: %s, length: %d\n", file.Name, file.Length)
+   }
+
+.. _golang-download-files:
+
+Download Files
+~~~~~~~~~~~~~~
+
+You can download a GridFS file in one of the following ways:
+
+- Use the ``DowloadToStream()`` method to download a file to an output stream.
+- Use the ``OpenDownloadStream()`` method to open an input stream.
+
+Download a File to an Output Stream
+```````````````````````````````````
+
+You can download a file in a GridFS bucket directly to an output stream using the
+``DownloadToStream()`` method. ``DownloadToStream()`` takes a file ID and an
+``io.Writer`` as parameters. The method downloads the file with the specified
+file ID and writes to the ``io.Writer``.
+
+The following example downloads a file and writes to a file buffer:
+
+.. code-block:: go
+
+   id, err = primitive.ObjectIDFromHex("62f7bd54a6e4452da13b3e88")
+   fileBuffer := bytes.NewBuffer(nil)
+   if _, err := bucket.DownloadToStream(id, fileBuffer); err != nil {
+      panic(err)
+   }
+
+Download a File to an Input Stream
+``````````````````````````````````
+
+You can download a file in a GridFS bucket to memory with an input stream using
+the ``OpenDownloadStream()`` method. ``OpenDownloadStream()`` takes a file ID as
+a parameter and returns an input stream from which you can read the file.
+
+The following example downloads a file into memory and reads its contents:
+
+.. code-block:: go
+
+   id, err = primitive.ObjectIDFromHex("62f7bd54a6e4452da13b3e88")
+   downloadStream, err := bucket.OpenDownloadStream(id)
+   if err != nil {
+      panic(err)
+   }
+
+   fileBuffer := bytes.NewBuffer(nil)
+   if _, err := downloadStream.Read(fileBuffer); err != nil {
+      panic(err)
+   }
```

Comment:
```suggestion
   fileBytes := make([]byte, 1024)
   if _, err := downloadStream.Read(fileBytes); err != nil {
      panic(err)
   }
```

This example does not compile:

```go
docs/main.go:36:34: cannot use fileBuffer (type *bytes.Buffer) as type []byte in argument to downloadStream.Read
```

For future reference, consider all examples in the `*.go` files in the `code-snippets` directory. That may make it easier to check they compile.

Reply:
Fixed this example